toolkit/governed-map/primitives/src/lib.rs

```diff
@@ -50,3 +69,167 @@ impl IsFatalError for InherentError {
+
+/// Error type returned when creation of [GovernedMapInherentDataProvider] fails.
+#[cfg(feature = "std")]
+#[derive(Debug, thiserror::Error)]
+pub enum InherentProviderCreationError {
+	#[error("Runtime API call failed: {0}")]
+	ApiError(#[from] sp_api::ApiError),
+	#[error("Data source call failed: {0}")]
+	DataSourceError(Box<dyn std::error::Error + Send + Sync>),
+	#[error("Unsupported pallet version {0} (highest supported version: {1})")]
+	UnsupportedPalletVersion(u32, u32),
+}
+
+#[cfg(feature = "std")]
+impl GovernedMapInherentDataProvider {
+	/// Creates a new [GovernedMapInherentDataProvider] with reference to the passed Cardano block hash.
+	///
+	/// This function is version aware and will return:
+	/// - an inert IDP if the pallet is not present in the runtime
+	/// - an inert IDP if the pallet does not have main chain scripts configured yet
+	/// - an IDP compatible with the pallet version signalled through the runtime API
+	///
+	/// Parameters:
+	/// - `client`: runtime client capable of providing [GovernedMapApi] runtime API
+	/// - `parent_hash`: parent hash of the current block
+	/// - `mc_hash`: Cardano block hash referenced by the current block
+	/// - `data_source`: data source implementing [GovernedMapDataSource]
+	pub async fn new<T, Block>(
+		client: &T,
+		parent_hash: Block::Hash,
+		mc_hash: McBlockHash,
+		data_source: &(dyn GovernedMapDataSource + Send + Sync),
+	) -> Result<Self, InherentProviderCreationError>
+	where
+		Block: sp_runtime::traits::Block,
+		T: ProvideRuntimeApi<Block> + Send + Sync,
+		T::Api: GovernedMapApi<Block>,
+	{
+		let api = client.runtime_api();
+		if !api.has_api::<dyn GovernedMapApi<Block>>(parent_hash)? {
+			log::info!("💤 Skipping Governed Map observation. Pallet not detected in the runtime.");
+			return Ok(Self::Inert);
+		}
+
+		match api.get_pallet_version(parent_hash)? {
+			1 => Self::new_v1(api.deref(), parent_hash, mc_hash, data_source).await,
+			unsupported_version => {
+				Err(InherentProviderCreationError::UnsupportedPalletVersion(unsupported_version, 1))
+			},
+		}
+	}
+
+	async fn new_v1<Api, Block>(
+		api: &Api,
+		parent_hash: Block::Hash,
+		mc_hash: McBlockHash,
+		data_source: &(dyn GovernedMapDataSource + Send + Sync),
+	) -> Result<Self, InherentProviderCreationError>
+	where
+		Block: sp_runtime::traits::Block,
+		Api: GovernedMapApi<Block>,
+	{
+		let Some(main_chain_script) = api.get_main_chain_scripts(parent_hash)? else {
+			log::info!("💤 Skipping Governed Map observation. Main chain scripts not set yet.");
+			return Ok(Self::Inert);
+		};
+
+		let entries_in_storage = api.get_stored_mappings(parent_hash)?;
+		let current_entries = data_source
+			.get_current_mappings(mc_hash, main_chain_script)
+			.await
+			.map_err(InherentProviderCreationError::DataSourceError)?;
+
+		let mut changes: ChangesV1 = ChangesV1::new();
+
+		for (key, value) in current_entries.iter() {
+			if !entries_in_storage.contains_key(key) || entries_in_storage[key] != *value {
```

Comment:
```suggestion
			if entries_in_storage.get(key) != Some(value) {
```
to avoid double lookup for the same key

Reply:
Good idea, thanks